### ISSUE TITLE
can no longer get changes from replication slot

### ISSUE THREAD

**Issue by Medzon888** [patroni/patroni#3150]
### What happened?

We have a patroni cluster with 3 nodes. They working fine. We have two logical replication slots on primary wich is used by CDC. For avoid problems with switchover's we perform a static logical slots in our configuration. And there is a problem.
ASYNC replication node (02) just restarted in cycle because of cant get changes from replication slot. 
```

+ Cluster: main (************************) -----------------------------------------------------------------------+--------------+-----------+----+-----------+
| Member                         | Host                                                                      | Role         | State     | TL | Lag in MB |
+--------------------------------+---------------------------------------------------------------------------+--------------+-----------+----+-----------+
| p-tms-picking-patroni-01-db-01 | p-tms-picking-patroni-01-db-01.******************************:5435        | Sync Standby | streaming |  1 |         0 |
| p-tms-picking-patroni-01-db-02 | p-tms-picking-patroni-01-db-02.**************:5435                        | Replica      | streaming |  1 |         0 |
| p-tms-picking-patroni-01-db-03 | p-tms-picking-patroni-01-db-03.**************:5435                        | Leader       | running   |  1 |           |
+--------------------------------+---------------------------------------------------------------------------+--------------+-----------+----+-----------+

```

Cycle in postgres log file.

Interesting that p-tms-picking-patroni-01-db-01 didn't have this problems and when I make 02 as sync replica then problem still in place. Trying to re-init: no luck.

We have 4 similar clusters in total with different applications and databases where we have logical slots for CDC and 2 of them have this issue when two others is clear. Whats wrong?

### How can we reproduce it (as minimally and precisely as possible)?

As this is not reproduced in every environment so I don't know...
In basic this is standart patroni cluster with 3 nodes, static slots and configuration below. There can be logical replicaion slots for CDC or maybe traditional logical replication to another PG cluster.

### What did you expect to happen?

Working without restarts and confilcts with recovery.

### Patroni/PostgreSQL/DCS version

-   Patroni version: 4.0.1
-   PostgreSQL version: 16.4
-   DCS (and its version): ETCD 3.5.7

### Patroni configuration file

```yaml
scope: main
namespace: /service/
name: p-tms-picking-patroni-01-db-03

log:
  level: DEBUG
  format: '%(asctime)s %(levelname)s: %(message)s'
  max_queue_size: 1000
  dir: /data/pg_data/logs/patroni
  file_num: 4
  file_size: 25000000
  loggers:
    patroni.postmaster: DEBUG
    urllib3: DEBUG

restapi:
  connect_address: p-tms-picking-patroni-01-db-03.***:8008
  listen: 0.0.0.0:8008
  authentication:
    username: patroni
    password: ***



etcd3:
  hosts: ***:2379,***:2379,***:2379
  username: root
  password: ***

bootstrap:
  dcs:
    ttl: 30
    loop_wait: 10
    retry_timeout: 10
    maximum_lag_on_failover: 1048576
    master_start_timeout: 300
    synchronous_mode: true
    synchronous_mode_strict: false
    check_timeline: true
    postgresql:
      use_pg_rewind: false
      use_slots: true
      parameters:
        max_connections: 300
        max_locks_per_transaction: 64
        max_worker_processes: 8
        max_prepared_transactions: 0
        wal_level: logical
        wal_log_hints: on
        track_commit_timestamp: off
        max_wal_senders: 10
        max_replication_slots: 10
  initdb:
    - encoding: UTF8
    - data-checksums
    - waldir: /data/pg_wal/postgres
  users:
    postgres:
      password: ***
    replica:
      password: ***
      options:
        - replication

postgresql:
  use_unix_socket: true
  listen: 0.0.0.0:5435
  connect_address: p-tms-picking-patroni-01-db-03.***:5435
  data_dir: /data/pg_data/postgres
  bin_dir: /usr/lib/postgresql/16/bin
  config_dir: /etc/postgresql/16/main
  pgpass: /var/lib/postgresql/.pgpass
  authentication:
    superuser:
      username: postgres
      password: ***
    replication:
      username: replica
      password: ***
  create_replica_methods:
    - basebackup
  basebackup:
    waldir: /data/pg_wal/postgres
  parameters:
    unix_socket_directories: /var/run/postgresql
    shared_preload_libraries: pg_stat_statements,auto_explain
    archive_mode: on
    archive_command: wal-g wal-push %p
  pg_hba:
    - local all all trust
    - host all all 127.0.0.1/32 trust
    - host all all 0.0.0.0/0 scram-sha-256
    - host replication replica 127.0.0.1/32 scram-sha-256
    - host replication replica ***8/32 scram-sha-256
    - host replication replica ***7/32 scram-sha-256
    - host replication replica ***/32 scram-sha-256
  pg_ctl_timeout: 60

watchdog:
  mode: automatic
  device: /dev/watchdog
  safety_margin: 5

tags:
    nofailover: false
    noloadbalance: false
    clonefrom: false
    nosync: false
```

### patronictl show-config

```yaml
check_timeline: true
loop_wait: 10
master_start_timeout: 300
maximum_lag_on_failover: 1048576
postgresql:
  parameters:
    max_connections: 300
    max_locks_per_transaction: 64
    max_prepared_transactions: 0
    max_replication_slots: 10
    max_wal_senders: 10
    max_worker_processes: 8
    track_commit_timestamp: false
    wal_level: logical
    wal_log_hints: true
  use_pg_rewind: false
  use_slots: true
retry_timeout: 10
slots:
  picking_pdpkfk_slot:
    database: picking
    plugin: pgoutput
    type: logical
  picking_rtdwh_pdpkfk_slot:
    database: picking
    plugin: pgoutput
    type: logical
synchronous_mode: true
synchronous_mode_strict: false
ttl: 30
```

### Patroni log files

```shell
Nothing special
```

### PostgreSQL log files

```shell
2024-09-06 12:37:01.812 MSK [2770070] [0] []: user=,db=,app=,client= LOG:  starting PostgreSQL 16.4 (Ubuntu 16.4-1.pgdg24.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 13.2.0-23ubuntu4) 13.2.0, 64-bit
2024-09-06 12:37:01.812 MSK [2770070] [0] []: user=,db=,app=,client= LOG:  listening on IPv4 address "0.0.0.0", port 5435
2024-09-06 12:37:01.832 MSK [2770070] [0] []: user=,db=,app=,client= LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5435"
2024-09-06 12:37:01.866 MSK [2770077] [0] []: user=,db=,app=,client= LOG:  database system was shut down in recovery at 2024-09-06 12:37:01 MSK
2024-09-06 12:37:01.866 MSK [2770077] [0] []: user=,db=,app=,client= LOG:  entering standby mode
2024-09-06 12:37:01.897 MSK [2770077] [0] [1/0]: user=,db=,app=,client= LOG:  redo starts at 149/A20FFE0
.......
2024-09-06 12:37:24.905 MSK [2770077] [0] [1/0]: user=,db=,app=,client= LOG:  invalidating obsolete replication slot "picking_rtdwh_pdpkfk_slot"
2024-09-06 12:37:24.905 MSK [2770077] [0] [1/0]: user=,db=,app=,client= DETAIL:  The slot conflicted with xid horizon 5642254.
2024-09-06 12:37:24.905 MSK [2770077] [0] [1/0]: user=,db=,app=,client= CONTEXT:  WAL redo at 149/2B52F198 for Heap2/PRUNE: snapshotConflictHorizon: 5642254, nredirected: 1, ndead: 0, nunused: 1, redirected: [3->4], dead: [], unused: [2]; blkref #0: rel 1663/255842/2619, blk 39
2024-09-06 12:37:24.949 MSK [2770293] [0] []: user=[unknown],db=[unknown],app=[unknown],client=[local] LOG:  connection received: host=[local]
2024-09-06 12:37:24.950 MSK [2770293] [0] [4/1]: user=postgres,db=picking,app=[unknown],client=[local] LOG:  connection authorized: user=postgres database=picking application_name=Patroni
2024-09-06 12:37:25.043 MSK [2770293] [0] [4/2]: user=postgres,db=picking,app=Patroni,client=[local] ERROR:  can no longer get changes from replication slot "picking_rtdwh_pdpkfk_slot"
2024-09-06 12:37:25.043 MSK [2770293] [0] [4/2]: user=postgres,db=picking,app=Patroni,client=[local] DETAIL:  This slot has been invalidated because it was conflicting with recovery.
2024-09-06 12:37:25.043 MSK [2770293] [0] [4/2]: user=postgres,db=picking,app=Patroni,client=[local] STATEMENT:  SELECT pg_catalog.pg_replication_slot_advance('picking_rtdwh_pdpkfk_slot', '149/2B378AE0')
2024-09-06 12:37:25.044 MSK [2770293] [0] [4/3]: user=postgres,db=picking,app=Patroni,client=[local] LOG:  starting logical decoding for slot "picking_pdpkfk_slot"
2024-09-06 12:37:25.044 MSK [2770293] [0] [4/3]: user=postgres,db=picking,app=Patroni,client=[local] DETAIL:  Streaming transactions committing after 149/2ADB2660, reading WAL from 149/2A7B5388.
2024-09-06 12:37:25.044 MSK [2770293] [0] [4/3]: user=postgres,db=picking,app=Patroni,client=[local] STATEMENT:  SELECT pg_catalog.pg_replication_slot_advance('picking_pdpkfk_slot', '149/2B3FC2D0')
2024-09-06 12:37:25.046 MSK [2770293] [0] [4/3]: user=postgres,db=picking,app=Patroni,client=[local] LOG:  logical decoding found consistent point at 149/2A7B5388
2024-09-06 12:37:25.046 MSK [2770293] [0] [4/3]: user=postgres,db=picking,app=Patroni,client=[local] DETAIL:  Logical decoding will begin using saved snapshot.
2024-09-06 12:37:25.046 MSK [2770293] [0] [4/3]: user=postgres,db=picking,app=Patroni,client=[local] STATEMENT:  SELECT pg_catalog.pg_replication_slot_advance('picking_pdpkfk_slot', '149/2B3FC2D0')
2024-09-06 12:37:25.145 MSK [2770293] [0] [4/0]: user=postgres,db=picking,app=Patroni,client=[local] LOG:  duration: 102.319 ms  statement: SELECT pg_catalog.pg_replication_slot_advance('picking_pdpkfk_slot', '149/2B3FC2D0')
....
2024-09-06 12:37:35.160 MSK [2770070] [0] []: user=,db=,app=,client= LOG:  received fast shutdown request
2024-09-06 12:37:35.163 MSK [2770070] [0] []: user=,db=,app=,client= LOG:  aborting any active transactions
2024-09-06 12:37:35.163 MSK [2770111] [0] [3/0]: user=postgres,db=postgres,app=Patroni restapi,client=[local] FATAL:  terminating connection due to administrator command
2024-09-06 12:37:35.163 MSK [2770110] [0] [2/0]: user=postgres,db=postgres,app=Patroni heartbeat,client=[local] FATAL:  terminating connection due to administrator command
2024-09-06 12:37:35.163 MSK [2770106] [0] []: user=,db=,app=,client= FATAL:  terminating walreceiver process due to administrator command
2024-09-06 12:37:35.163 MSK [2770110] [0] []: user=postgres,db=postgres,app=Patroni heartbeat,client=[local] LOG:  disconnection: session time: 0:00:31.375 user=postgres database=postgres host=[local]
2024-09-06 12:37:35.163 MSK [2770111] [0] []: user=postgres,db=postgres,app=Patroni restapi,client=[local] LOG:  disconnection: session time: 0:00:31.371 user=postgres database=postgres host=[local]
2024-09-06 12:37:35.177 MSK [2770075] [0] []: user=,db=,app=,client= LOG:  shutting down
2024-09-06 12:37:35.272 MSK [2770070] [0] []: user=,db=,app=,client= LOG:  database system is shut down
2024-09-06 12:37:35.827 MSK [2770401] [0] []: user=,db=,app=,client= LOG:  starting PostgreSQL 16.4 (Ubuntu 16.4-1.pgdg24.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 13.2.0-23ubuntu4) 13.2.0, 64-bit
2024-09-06 12:37:35.827 MSK [2770401] [0] []: user=,db=,app=,client= LOG:  listening on IPv4 address "0.0.0.0", port 5435
2024-09-06 12:37:35.849 MSK [2770401] [0] []: user=,db=,app=,client= LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5435"
2024-09-06 12:37:35.879 MSK [2770406] [0] []: user=,db=,app=,client= LOG:  database system was shut down in recovery at 2024-09-06 12:37:35 MSK
2024-09-06 12:37:35.879 MSK [2770406] [0] []: user=,db=,app=,client= LOG:  entering standby mode
2024-09-06 12:37:35.911 MSK [2770406] [0] [1/0]: user=,db=,app=,client= LOG:  redo starts at 149/A20FFE0
```

### Have you tried to use GitHub issue search?

- [X] Yes

### Anything else we need to know?

Looks like there is similar issue but this is physical slots and our issue is different
https://github.com/patroni/patroni/issues/3146

**Comment by CyberDem0n** [comment]
> Whats wrong?

Logical replication slot is invalidated by postgres:
```
2024-09-06 12:37:24.905 MSK [2770077] [0] [1/0]: user=,db=,app=,client= LOG:  invalidating obsolete replication slot "picking_rtdwh_pdpkfk_slot"
2024-09-06 12:37:24.905 MSK [2770077] [0] [1/0]: user=,db=,app=,client= DETAIL:  The slot conflicted with xid horizon 5642254.
2024-09-06 12:37:24.905 MSK [2770077] [0] [1/0]: user=,db=,app=,client= CONTEXT:  WAL redo at 149/2B52F198 for Heap2/PRUNE: snapshotConflictHorizon: 5642254, nredirected: 1, ndead: 0, nunused: 1, redirected: [3->4], dead: [], unused: [2]; blkref #0: rel 1663/255842/2619, blk 39
```

After that it is not possible to use it anymore. The only thing Patroni can do - recreate replication slot by copying it from the primary. Increasing `max_standby_streaming_delay` may help.

**Comment by Medzon888** [comment]
Already tryed but no luck. Also hot_standby_feedback must prevent it

**Comment by CyberDem0n** [comment]
> Aslo stange that another node works fine isn't it?

It's not a fault of Patroni. On other nodes you have a different workload.
You have to figure out what causing this conflict to happen.

**Comment by Medzon888** [comment]
Also found that when I turn on static slots via edit-config then slots created only for async replica bu there is no entries for sync replica. e.g. for sync replica slots din't created. Sorry for missundestanding. There is only entires from the log file
```
2024-09-06 21:07:03.445 MSK [1272036] [0] []: user=,db=,app=,client= LOG:  received SIGHUP, reloading configuration files
2024-09-06 21:07:03.446 MSK [1272036] [0] []: user=,db=,app=,client= LOG:  parameter "hot_standby_feedback" changed to "on"
```

**Comment by Medzon888** [comment]
The following bug fix in 3.3.2:

Handle logical slots invalidation on a standby (Polina Bungina)
Since PG16 logical replication slots on a standby can be invalidated due to horizon: from now on, Patroni forces copy (i.e., recreation) of invalidated slots. 

It can be the same situation in 4.0.1?

**Comment by CyberDem0n** [comment]
> Also found that when I turn on static slots via edit-config then slots created only for async replica bu there is no entries for sync replica.

On sync replica it might take a bit longer to copy them, but eventually it should happen.

> The following bug fix in 3.3.2:

The bugfix is about catching exception raised by Postgres when we are trying to advance invalidated logical replication slot. If the slot is invalidated it can't be no longer used and the only we can do - reinitialize it by copying from the primary.

> It can be the same situation in 4.0.1?

What you call a "situation" is an expected behavior. Invalidated logical replication slots are useless. You need to understand your workload and figure out the reason why invalidation happens.

**Comment by Medzon888** [comment]
Looks like fix from 4.0.2 works. I was upgrade patroni one hour ago and  enable static logical slots back. As result i did not see any issues as rebooting slave with xmin horizon error or bloating master node. 
I will update the post tomorrow with deeper testing.

**Comment by Medzon888** [comment]
Also guess I found out why a few clusters rebooted in a cycle when I enable static logical slots. One of cluster slave have wal_status = lost for physical slots. Maybe can be a reason

**Comment by Medzon888** [comment]
Fix in 4.0.2 works fine. No any issues now. Thanks

**Comment by Medzon888** [comment]
Fix in 4.0.2 works fine. No any issues now. Thanks
